jt_notes.md

```diff
@@ -0,0 +1,127 @@
+**I [jonny] have made changes to the source code on a new branch (todos, improved comments, and a few changes to function/variable names etc)**. Where possible I have edited as I go, but am hoping for your input on resolving the queries in the todos. The more important conceptual queries are listed below. **I have also included a preliminary sketch of the architecture I have in mind**, although this is currently limited by my understanding of how the flask stuff works, particularly around multithreading, since that will have a big influence on what a correct and best way to handle all this in python will be. My biggest hangup is that I cannot see any connection between app.py and camera_pi.py (and in particular the frames() iterator). I therefore cannot see how the Flask app stuff is meant to work when running live on a RPi with the PiCam.
+
+------------------------------------ **Queries to Chas about the different modules ------------------------------------**
+
+cpi.py
+
+OpticalGater class that extends picamera.array.PiYUVAnalysis. It appears to maintain history, e.g. frame_history, period_history.
+
+**For chas:** my opinion is frame_history is really crying out for a class container, or at least an array with named columns (or a list-of-dictionaries, though I'm less keen on that). At present it requires the developer to keep track of the meaning of each column, and the format not easily extensible.
```

Comment:
Suggestion: I think a class could be overkill here - unless there are particular methods you're thinking to relocate into the class? Alternatives are pandas DataFrames or xarrays.

Reply:
I'm happy with whatever you think is best, and most python-like. In c++ the distinction between structs and classes is mostly just nomenclature, and a struct would be the natural way to do this. In python I know there are various solutions, and I don't have much of a view about which is most appropriate.

Reply:
Could [data classes](https://docs.python.org/3/library/dataclasses.html) maybe be of use here? They might also be a bit overkill. I'm not sure how the compare to DataFrames or xarrays in terms of speed/performance.

Reply:
That does sound like a useful thing for me to know about, which I will keep in mind for the future! Looks much like a fancy (in a good way) counterpart to a C++ struct in many ways.

In terms of what to do here, I think we are probably going for a metadata dictionary attached to PixelArray frame objects. i.e. these quantities will not be _separately_ stored _anywhere_ (*). Certainly I see sync-related properties such as 'phase' being included in there. There is an accessor function in pixelarray.py to pull out a 1D array for a given metadata key, when needed, and its performance turns out to be plenty good enough.

(*) I am aware that if we choose to maintain _very long_ histories for some reason, then we may need to just archive the metadata rather than the (large) raw pixel data.